### ISSUE TITLE
When the depth > 3, variable (invariant and coors_out) will be NaN

### ISSUE THREAD

**Issue by JunDecNo** [lucidrains/gotennet-pytorch#2]
I encountered this issue when using the example and setting the depth to 4.

**Comment by lucidrains** [comment]
@JunDecNo indeed it does

i threw in an extra norm in a place that makes sense, and stabilized it

do you want to perhaps send the author of this paper an email and see if he is missing normalization in the HTR?

**Comment by lucidrains** [comment]
this was not a well written paper.

**Comment by anine09** [comment]
@lucidrains Have you looked at the code for [this repository](https://github.com/knowledge-anonymous/review) ? Everything seems to work fine when I use it.

**Comment by lucidrains** [comment]
@anine09 oh nice, the code was released after all! why did I bother 🤔🤣

**Comment by lucidrains** [comment]
I'll reconcile the differences later this week and wrap up the project

**Comment by anine09** [comment]
After you answered my query about `lens` in version 0.2.4, I was also experiencing output as NaN, then I found this library, and recently I switched to their implementation, and right after I commented on this issue, my loss immediately turned into a loss, so please be careful! Good luck!🤣

**Comment by cenjc** [comment]
I give an explanation. High-degree steerable features can be understood as electron clouds. The inner product describes the direct phase relationship between two electron clouds. The meaning of the modulus component is relatively weak, so the modulus component can be completely separated.

In addition, Nan may be due to multi-channel reasons. This has been corrected by using the Frobenius norm in GMN[1]. In addition, in section 3.4 of EquiformerV2[2], there is a similar correction method in the HEGNN[3] code (https://github.com/GLAD-RUC/HEGNN/blob/main/models/HEGNN.py#L94). However, the method of directly correcting high-degree steerable features used in the GotenNet implementation mentioned by anine09 is also very interesting.

Finally, I also recommend our HEGNN project (https://github.com/GLAD-RUC/HEGNN). You can consider using e3nn.o3.FullyConnectedTensorProduct to implement multi-channel high-degree steerable features (e.g. Eqs. (10)-(11) in the current version of GotenNet paper).

[1] Equivariant Graph Mechanics Networks with Constraints
[2] EquiformerV2: Improved Equivariant Transformer for Scaling to Higher-Degree Representations
[3] Are High-Degree Representations Really Unnecessary in Equivarinat Graph Neural Networks?

**Comment by kdmayer** [comment]
For other people who face this issue, casting the model and inputs to .double() allows training at higher model depths without experiencing the NaN issue.

**Comment by sarpaykent** [comment]
Hi @lucidrains,

Firstly, thank you very much for your interest in GotenNet and for putting together this implementation so swiftly!

I wanted to let you and the community know that we've recently released the official v1 reference [implementation](https://github.com/sarpaykent/GotenNet/) of GotenNet. This version is tightly aligned with the methodology described in the paper.

Regarding the NaN instability issue reported by @JunDecNo: In our reference implementation, we did not encounter NaN issues during training on the QM9, rMD17, and MD22 datasets, even with deep models (12 layers). The only dataset where stability required specific attention was Molecule3D, for which we utilized [TensorLayerNorm](https://github.com/sarpaykent/GotenNet/blob/main/gotennet/models/components/layers.py#L1497).

Hopefully, the official release helps clarify the implementation details. Thanks again for your contribution!

**Comment by lucidrains** [comment]
also, i think the stability issues may have to do with https://github.com/lucidrains/gotennet-pytorch/issues/4

**Comment by sarpaykent** [comment]
Thanks @lucidrains for linking the repo!

Yes, absolutely, I'd be interested to chat on Signal about the direction of geometric learning. And regarding stability, #4 indeed explain the stability differences encountered earlier.

**Comment by JunDecNo** [comment]
I apologize for the interruption. I compared gotennet-pytorch to the official implementation, the reason for the NaN values is likely due to the absence of an activation function. In the official implementation [L325](https://github.com/sarpaykent/GotenNet/blob/7134e2d327c43d8ed9de562fbfeeb59c92a8b627/src/models/representation/gotennet.py#L325), 
```python
if self.update_info["gated"] == "gatedt":
	w_dot = torch.tanh(w_dot)
elif self.update_info["gated"] == "gated":
	w_dot = torch.sigmoid(w_dot)
elif self.update_info["gated"] == "act":
	w_dot = self.activation(w_dot)
```
Additionally, combining [L332](https://github.com/sarpaykent/GotenNet/blob/7134e2d327c43d8ed9de562fbfeeb59c92a8b627/src/models/representation/gotennet.py#L332) and [L227](https://github.com/sarpaykent/GotenNet/blob/7134e2d327c43d8ed9de562fbfeeb59c92a8b627/src/models/representation/gotennet.py#L227), it should require adding residual connections #4, similar to:
```python
t_ij_residual = t_ij
...
edge_proj_out = self.edge_proj(w_ij)
edge_proj_out = torch.sigmoid(edge_proj_out)

residue_update_out = self.residue_update(t_ij)
delta_t_ij = edge_proj_out * residue_update_out
return t_ij_residual + delta_t_ij
```
I still have a point of confusion: in the official implementation, the `edge_updater` is executed after `propagate`. Should the execution of `htr` in gotennet-pytorch be adjusted accordingly? (`propagate` in [L206](https://github.com/sarpaykent/GotenNet/blob/7134e2d327c43d8ed9de562fbfeeb59c92a8b627/src/models/representation/gotennet.py#L206), `edge_updater` in [L227](https://github.com/sarpaykent/GotenNet/blob/7134e2d327c43d8ed9de562fbfeeb59c92a8b627/src/models/representation/gotennet.py#L227)) This might require the assistance of @sarpaykent.

**Comment by lucidrains** [comment]
@JunDecNo thanks Jun, made the change
